### DISCUSSION TITLE
The Rust Programming Language

### DISCUSSION THREAD

**Discussion by lxrst** [immich-app/immich#1809] [General]
This probably won't go well, but here goes.

Crates of interest:
[`image`](https://crates.io/crates/image)
[`kamadak-exif`](https://crates.io/crates/kamadak-exif)

**Comment by lxrst** [comment]
Ofc completely rewriting Immich in Rust is too much, but slowly oxidizing parts may benefit the project.
EDIT: Would like to add that the image crate is *awesome*.

**Reply by alextran1502** [reply]
Jokes aside, I started writing Immich in Rust, but the need for a uniform framework that leads to the reason of going with JavaScript/TypeScript to attract more contributors. 

I appreciate Rust no less than you are; it is a beautiful language and super performance. But the trade-off is huge regarding the community and ease of contribution. 

The use case of Immich is self-hosted. I don't see many instances that hosted more than ten people. With that, JS/TS outweighs the trade-off of ease of use vs performance.

**Comment by zackpollard** [comment]
https://user-images.githubusercontent.com/1313578/220212873-864b6fa3-ac27-441c-bf32-5f60aaeaefe2.mp4

**Reply by Zenexer** [reply]
Really, the next step should be C#, as that seems like the natural progression from TypeScript.

All in favor of a C# rewrite?

**Comment by giripriyadarshan** [comment]
will switching to rust make the server performant enough to add encryption of media feature? It was mentioned in #1150 that it is a performance issue

**Reply by uhthomas** [reply]
Please read #450, performance is not the issue and Immich will likely not be rewritten in rust.

**Reply by lxrst** [reply]
As Thomas said, there's more to it than just performance. Yes, Rust could improve performance enough to do something like this, but encryption would impact Immich's functionality too much to be worth it.
Also, rewriting is a huge task and isn't going to happen unless somebody has enough time, know-how, and persistence. I doubt Immich itself is going to be rewritten, but an alternate server implementation wouldn't be impossible.
Surprised to see activity nearly a year later.
